main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -9,94 +9,242 @@ package sbt
+                  console.appendLog(Level.Error, "sbt server disconnected")
+                  exitClean.set(false)
+                }
+              }
+              stdinBytes.offer(-1)
+              Option(inputThread.get).foreach(_.close())
+              Option(interactiveThread.get).foreach(_.interrupt)
+            case "readInput" =>
+            case _           => self.onNotification(msg)
+          }
+        }
+        override def onRequest(msg: JsonRpcRequestMessage): Unit = self.onRequest(msg)
+        override def onResponse(msg: JsonRpcResponseMessage): Unit = self.onResponse(msg)
+        override def onShutdown(): Unit = {
+          if (exitClean.get != false) exitClean.set(!running.get)
+          running.set(false)
+          Option(interactiveThread.get).foreach(_.interrupt())
+        }
       }
+      // initiate handshake
+      val execId = UUID.randomUUID.toString
+      val initCommand = InitCommand(tkn, Option(execId), Some(true))
+      conn.sendString(Serialization.serializeCommandAsJsonMessage(initCommand))
+      connectionHolder.set(conn)
+      conn
+    } catch {
+      case e: ConnectionRefusedException if retry =>
+        if (Files.deleteIfExists(portfile.toPath)) init(prompt, retry = false)
+        else throw e
     }
-    // initiate handshake
-    val execId = UUID.randomUUID.toString
-    val initCommand = InitCommand(tkn, Option(execId))
-    conn.sendString(Serialization.serializeCommandAsJsonMessage(initCommand))
-    conn
-  }
 
   /**
    * Forks another instance of sbt in the background.
    * This instance must be shutdown explicitly via `sbt -client shutdown`
    */
-  def forkServer(portfile: File): Unit = {
-    console.appendLog(Level.Info, "server was not detected. starting an instance")
-    val args = List[String]()
-    val launchOpts = List("-Xms2048M", "-Xmx2048M", "-Xss2M")
-    val launcherJarString = sys.props.get("java.class.path") match {
-      case Some(cp) =>
-        cp.split(File.pathSeparator)
-          .toList
-          .headOption
-          .getOrElse(sys.error("launcher JAR classpath not found"))
-      case _ => sys.error("property java.class.path expected")
-    }
-    val cmd = "java" :: launchOpts ::: "-jar" :: launcherJarString :: args
-    // val cmd = "sbt"
-    val io = BasicIO(false, ProcessLogger(_ => ()))
-    val _ = Process(cmd, baseDirectory).run(io)
-    def waitForPortfile(n: Int): Unit =
-      if (portfile.exists) {
-        console.appendLog(Level.Info, "server found")
-      } else {
-        if (n <= 0) sys.error(s"timeout. $portfile is not found.")
-        else {
-          Thread.sleep(1000)
-          if ((n - 1) % 10 == 0) {
-            console.appendLog(Level.Info, "waiting for the server...")
-          }
-          waitForPortfile(n - 1)
+  def forkServer(portfile: File, log: Boolean): Unit = {
+    if (log) console.appendLog(Level.Info, "server was not detected. starting an instance")
```

Comment:
I wonder if we should also print a warning saying:

```
[info] starting sbt server in the background...
[warn] close the server with `sbtc shutdown` after using
```

Reply:
Yeah. I was thinking about something like that since people are used to quitting sbt with exit.

Reply:
https://github.com/sbt/sbt/pull/5620/commits/314d9a295d8b96a480243466ec11d68450ebb53e

Reply:
I didn't want to put sbtc in the instructions since users may alias the command. What do you think?

Reply:
Plus, I arbitrarily picked sbtc by fiat. Are we sure that's what we want to call it?